商品出品機能

sorairo580 opened the pull request:
# WHAT
商品出品機能の実装

#WHY
商品出品機能を実装するため

ログイン状態での出品ページへの遷移↓↓
https://gyazo.com/f1976c4d290ba4a1d9235ecdfd538f4d
価格入力時に販売手数料と販売利益の表示↓↓
https://gyazo.com/60bbf433047a54ba03ca9b30cf734923
正常に出品できた時の動画↓↓
https://gyazo.com/fa5677d0725fb60a2df0029d0defd149
入力不足の状態で出品ボタンを押したときの動画↓↓
https://gyazo.com/d989aba01354cca4cfb6e2f861354ba8
ログアウト状態で出品ボタンを押したときの動画↓↓
https://gyazo.com/8d222399387411bfd8de18330cb826b4
テスト結果の画像↓↓
https://gyazo.com/2d94841a1cf65afc6a09e7a3c442d00d
https://gyazo.com/5842068fd639f791604e3c1601fa4ed2

よろしくお願いいたします。

sorairo580 committed:
商品出品機能first commit

①itemテーブルを作成
②newアクションを作成
③Active Storageの導入
④itemモデルに画像ファイルの紐づけ
⑤画像保存許可のストロングパラメーター

sorairo580 committed:
ActiveHash導入

item用の各モデル作成
選択肢の記述

sorairo580 committed:
商品出品機能２

出品画面への遷移
Active Hashの実装・編集
出品価格の制限

sorairo580 committed:
エラーメッセージのエラー解消

そのほか商品化価格範囲のバリデーション

sorairo580 committed:
未ログイン時はログイン画面へ遷移

sorairo580 committed:
出品機能３

createアクションの修正
JavaScriptによる手数料と利益の表示
バリデーションにimageを追加
誤字（contition→condition）の修正
itemテーブルへuser_idカラムの追加
rubocopの使用

sorairo580 committed:
itemモデルのテストコード

TAKADA-yuma commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,33 @@
+  before_action :move_to_sessions_new, except: [:index, :show]
+
   def index
+    @items = Item.order('created_at DESC')
```
5行目

こちらの記述は今後の機能で実装予定のため、今はコメントアウトしておきましょう。

TAKADA-yuma commented on app/javascript/channels/index.js:
```diff
@@ -3,3 +3,22 @@
+      profitSpan.textContent = profit;
+    } 
+  })
+})
```
7〜24行目
これらのJavascriptは別にJavascriptのファイルを作りましょう。

TAKADA-yuma commented on app/javascript/channels/index.js:
```diff
@@ -3,3 +3,22 @@
+      let taxPrice = inputValue * 0.1; // 税金を計算
+      taxSpan.textContent = taxPrice;
+      let profit = inputValue - taxPrice; // 税金を計算
+      profitSpan.textContent = profit;
```
現状、商品の値段が少数を含んでいる場合にそのまま計算されてしまいます。
小数点以下は切り捨てられるようなメソッドを用いましょう。

TAKADA-yuma commented on app/models/item.rb:
```diff
@@ -1,2 +1,23 @@
+
+  validates :image,            presence: true
+  validates :item_name,        presence: true
+  validates :price,            presence: true, format: { with: /\A[0-9]+\z/ },
```
8行目
金額のバリデーション設置ありがとうございました。中身をもう一度確認してみましょう。

**■理由**
```ruby
・{ with: /\A[0-9]+\z/ } integer型に正規表現は適用ができない為


数値に対してバリデーションをかける場合、numericalityがおすすめです。
```

**■参考**
```ruby
コース：最終課題
タイトル：最終課題補足カリキュラム〜1つのフォームから複数のテーブルに情報を保存するアプリを作成しよう〜

バリデーションを設定しましょう
```

TAKADA-yuma commented on app/models/item.rb:
```diff
@@ -1,2 +1,23 @@
+  validates :image,            presence: true
+  validates :item_name,        presence: true
+  validates :price,            presence: true, format: { with: /\A[0-9]+\z/ },
+                               numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999_999, message: 'is invalid' }
```
9行目
金額のバリデーション設置ありがとうございました。中身をもう一度確認してみましょう。

現状、少数を含む数字も保存されてしまう可能性があります。
整数のみ保存可能なようにバリデーションの中身を修正しましょう

TAKADA-yuma commented on config/routes.rb:
```diff
@@ -1,4 +1,6 @@
   devise_for :users
   root to: 'items#index'
+  resources :items
+  post 'items', to: 'items#create'
```
5行目
こちらの記述は4行目に指定されているため不要になります。

TAKADA-yuma commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,71 @@
+        @item.price = 100
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is invalid')
+      end
```
以下の異常系テストも実施しましょう。
```
・価格に半角数字以外が含まれている場合は出品できない（※半角数字以外が一文字でも含まれていれば良い）
・価格が300円未満では出品できない
・価格が9_999_999円を超えると出品できない
・userが紐付いていなければ出品できない
```

TAKADA-yuma reviewed (commented):
レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

修正点を追加でコメントさせて頂きましたので、ご確認をお願い致します。

```ruby
【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

sorairo580 committed:
出品機能の修正

・未実施項目のコメントアウト
・JavaScriptファイルの分割
・金額のバリデーション修正（integer型、小数点以下切り捨て）
・重複した記述の削除
・異常系テスト４件追加

sorairo580 committed:
rubocopによる修正

sorairo580 committed:
販売履歴の小数点以下切り捨て

sorairo580 commented:
ご指導ありがとうございます。
出品機能の修正を行いました。

・未実施項目のコメントアウト
・JavaScriptファイルの分割及び金額のバリデーション修正（integer型、小数点以下切り捨て）（画像および動画↓↓）
https://gyazo.com/1edfa48919bf4d5cbf5f3b1858af3edd
https://gyazo.com/00d50d7f3a86314004be7a4132df2ed9
・重複した記述の削除
・異常系テスト４件追加（画像↓↓）
https://gyazo.com/9698625e1e9e46a367e1ca3b6a5437d8
https://gyazo.com/fcea30dd7edfd3caaf6f96117b74499e
https://gyazo.com/64938c0093e5cf5bc9e848bcbcf8c52a

TAKADA-yuma reviewed (commented):
レビュー依頼ありがとうございます！
LGTMです！
商品出品機能の実装、お疲れ様でした！

この後の流れの確認です。
①Git上でマージ
②ローカルのmasterブランチへ切り替え
③Fetch Originしプル

そのあと、新しいブランチを作成し、次の実装に取り組んでください。
引き続き、実装頑張っていきましょう。
応援しております！
